fe/fe-core/src/main/java/com/starrocks/StarRocksFE.java

```diff
@@ -302,4 +307,38 @@ private static boolean createAndLockPidFile(String pidFilePath) {
 
         return false;
     }
+
+    // NOTE: To avoid dead lock
+    //      1. never call System.exit in shutdownHook
+    //      2. shutdownHook cannot have lock conflict with the function calling System.exit
+    private static void addShutdownHook() {
```

Comment:
If someone finds FE is stuck in line 337 after being killed, he may kill FE once again. Better add a mechanism to avoid shutdown hook called multiple times, for example, add an independent lock to `flushVLSNMapping()` and set `replicatedEnvironment` null after flush.

Reply:
shutdown hook can only be call once, this is guaranteed by jdk.